### ISSUE TITLE
Tracking/Logging of validation loss

### ISSUE THREAD

**Issue by MatthiasEg** [RUCAIBox/RecBole#1831]
Hi, thank you for creating RecBole!

I was wondering if there's a way to enable tracking of the validation loss in addition to the train loss? This could be very helpful in order to determine overfitting/underfitting.

Thank you.

**Comment by MatthiasEg** [comment]
Not sure, but using LogLoss as a metric could maybe do the trick? (when using cross entropy as loss at least)

**Comment by Sherry-XLL** [comment]
Hello @MatthiasEg, thanks for your attention to RecBole!

I regret to tell you that in the current implementation of RecBole, we have not calculated the loss of the validation set. We only evaluate the results on the validation set, using the designated evaluation metric as a condition for early stopping. It is also a common practice in the field of recommender systems.

In terms of `LogLoss`, it is a value-based metric (also known as logistic loss or cross-entropy loss). `LogLoss` is used to evaluate the probabilistic output of the two-class classifier for context-aware recommendation. which has nothing to do with validation loss.

If you need it, you can simulate the training loss in the trainer's `fit` function to obtain the validation loss. We will also consider adding the feature of validation loss in future development. Thanks again for your feedback and suggestions.

**Comment by Sherry-XLL** [comment]
Hello @MatthiasEg! 

If we want to add evaluation loss, we can just modify the code for calculating training loss.

As for the `_train_epoch` function in [RecBole](https://github.com/RUCAIBox/RecBole/tree/master)/[recbole](https://github.com/RUCAIBox/RecBole/tree/master/recbole)/[trainer](https://github.com/RUCAIBox/RecBole/tree/master/recbole/trainer)/trainer.py, we calculate the training loss as follows:

https://github.com/RUCAIBox/RecBole/blob/2accad566b38bc009855c9ac14a3816a81a5eb1e/recbole/trainer/trainer.py#L201-L215

For evaluation loss, we can follow the code in `_train_epoch`, and make three modifications:
-  change `self.model.train()` to `self.model.eval()` in https://github.com/RUCAIBox/RecBole/blob/2accad566b38bc009855c9ac14a3816a81a5eb1e/recbole/trainer/trainer.py#L216
- change `train_data` to `valid_data` in https://github.com/RUCAIBox/RecBole/blob/2accad566b38bc009855c9ac14a3816a81a5eb1e/recbole/trainer/trainer.py#L219
- delete the process of back propagation in https://github.com/RUCAIBox/RecBole/blob/2accad566b38bc009855c9ac14a3816a81a5eb1e/recbole/trainer/trainer.py#L258-L267

**Comment by MatthiasEg** [comment]
@Sherry-XLL  Thanks for the heads-up!
It works with an additional slight change to access the actual val interaction from the tuple:

```python
for batch_idx, interaction in enumerate(iter_data):
    if is_val:
        interaction = interaction[0]
```

**Comment by Sherry-XLL** [comment]
Hi @MatthiasEg, I am glad to see that you have found a solution to your problem. 

Thanks again for your feedback and suggestions.

Feel free to contact us if you have further questions.

**Comment by MatthiasEg** [comment]
Hi @Sherry-XLL, could you consider looking more closely into this? I would argue that tracking the validation loss is very important and informative in terms of determining the level of overfitting/underfitting. Or how else do you suggest doing this using RecBole? 

When it comes to the messy implementation above, I discovered two things:

- It uses a lot more VRAM, hence is not really viable
- The validation loss is on another scale than the train loss, making it harder to visually compare train and val loss (although I'm not sure if that can be addressed)

Thank you.

**Comment by Sherry-XLL** [comment]
Hello @MatthiasEg, tracking the loss of validation sets is not the only way to avoid overfitting. 

RecBole has provided the validation score to avoid overfitting, which means that we will evaluate the performance of the currently trained model's metrics on the validation set, such as Recall or NDCG. If metrics in the validation set do not improve after exceeding the `stopping_step`, we will stop training by early stopping. This is also a common strategy academic papers on recommender systems. 

I suggest that you do not need to implement validation set loss on your own, but instead use the validation set indicator `valid_metric` that RecBole has already implemented as an early stop indicator to meet regular needs.
